File: latest/week3/workshop7_m7/W07_M07_demo.py
"""W07-M07: Case Study — Netflix Data Story — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
from matplotlib.dates import DateFormatter
from matplotlib.gridspec import GridSpec
os.makedirs("output", exist_ok=True)

# Clean defaults
plt.rcParams.update({"axes.spines.top":False,"axes.spines.right":False,
    "axes.edgecolor":"#EEE","axes.labelcolor":"#888","xtick.color":"#888",
    "ytick.color":"#888","grid.alpha":0,"font.size":10,"axes.titleweight":"bold"})

nf = pd.read_csv("netflix.csv")
nf["date_added"] = pd.to_datetime(nf["date_added"].str.strip(), errors="coerce")
nf["year_added"] = nf["date_added"].dt.year
nf["primary_country"] = nf["country"].str.split(",").str[0].str.strip()
yearly = nf.query("2015<=year_added<=2021").groupby(["year_added","type"]).size().reset_index(name="n")
movies = yearly.query("type=='Movie'"); tvshows = yearly.query("type=='TV Show'")
peak_n = movies.n.max(); latest_n = movies.iloc[-1]["n"]
decline = round((1 - latest_n / peak_n) * 100)

# ═══════════════════════════════════════════════════════
# Build all 8 evidence slides
# ═══════════════════════════════════════════════════════

# S2: Context — cumulative area
fig, ax = plt.subplots(figsize=(10, 5.5))
cum = nf.dropna(subset=["date_added"]).set_index("date_added").resample("ME").size().cumsum()
ax.fill_between(cum.index, cum.values, alpha=0.15, color="#1565C0")
ax.plot(cum.index, cum.values, color="#1565C0", lw=1.5)
ax.xaxis.set_major_formatter(DateFormatter("%Y"))
ax.set_title("Netflix built a catalog of 8,800+ titles over a decade", fontsize=13)
ax.set_ylabel("Cumulative Titles")
plt.tight_layout(); plt.savefig("output/slide02_context.png", dpi=300); plt.close()

# S3: Complication — movie decline (grey+accent)
fig, ax = plt.subplots(figsize=(10, 5.5))
ax.plot(tvshows.year_added, tvshows.n, "-o", color="#DDD", lw=0.8, markersize=3)
ax.plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=2.5, markersize=6)
ax.text(2021.15, latest_n, " Movies", fontsize=9, fontweight="bold", color="#1565C0", va="center")
ax.text(2021.15, tvshows.iloc[-1]["n"], " TV Shows", fontsize=8, color="#BBB", va="center")
ax.annotate(f"–{decline}%\nfrom peak", xy=(2019.5, (peak_n+latest_n)/2),
    fontsize=11, fontweight="bold", color="#E53935", ha="center",
    bbox=dict(boxstyle="round,pad=0.3", facecolor="white", edgecolor="#E53935"))
ax.set_xlim(2014.5, 2022.5)
ax.set_title(f"Movie additions declined {decline}% from their 2019 peak", fontsize=13)
ax.set_ylabel("Titles Added")
plt.tight_layout(); plt.savefig("output/slide03_complication.png", dpi=300); plt.close()

# S4: Contrast — indexed dual line
fig, ax = plt.subplots(figsize=(10, 5.5))
m_idx = movies.n.values / movies.n.values[0] * 100
t_idx = tvshows.n.values / tvshows.n.values[0] * 100
ax.plot(movies.year_added, m_idx, "-o", color="#1565C0", lw=1.5, markersize=5, label="Movie")
ax.plot(tvshows.year_added, t_idx, "-o", color="#E53935", lw=1.5, markersize=5, label="TV Show")
ax.axhline(100, ls=":", color="#888", lw=0.5)
ax.text(2021.15, m_idx[-1], " Movies", fontsize=8, fontweight="bold", color="#1565C0", va="center")
ax.text(2021.15, t_idx[-1], " TV Shows", fontsize=8, fontweight="bold", color="#E53935", va="center")
ax.set_xlim(2014.5, 2022.5)
ax.set_title("TV Shows grew 3× faster than Movies in relative terms", fontsize=13)
ax.set_ylabel("Index (2015 = 100)")
plt.tight_layout(); plt.savefig("output/slide04_contrast.png", dpi=300); plt.close()

# S5: Analysis — genre growth (accent bar)
genres = nf["listed_in"].str.split(",").explode().str.strip()
g17 = nf.query("year_added==2017")["listed_in"].str.split(",").explode().str.strip().value_counts()
g21 = nf.query("year_added==2021")["listed_in"].str.split(",").explode().str.strip().value_counts()
growth = ((g21 - g17) / g17.clip(lower=1) * 100).dropna().sort_values().tail(7)
fastest_genre = growth.index[-1]

fig, ax = plt.subplots(figsize=(10, 5.5))
colors = ["#2E7D32" if g == fastest_genre else "#BBDEFB" for g in growth.index]
ax.barh(growth.index, growth.values, color=colors, height=0.5)
for i, (g, v) in enumerate(growth.items()):
    ax.text(v+2, i, f"+{v:.0f}%", va="center", fontsize=8, fontweight="bold",
        color="#2E7D32" if g == fastest_genre else "#888")
ax.set_title(f"{fastest_genre} is the fastest-growing genre category", fontsize=13)
ax.set_xlabel("Growth % (2017 → 2021)")
plt.tight_layout(); plt.savefig("output/slide05_analysis.png", dpi=300); plt.close()

# S6: Deep Dive — India bump chart
cyr = (nf.dropna(subset=["primary_country"]).query("2016<=year_added<=2021")
    .groupby(["year_added","primary_country"]).size().reset_index(name="n"))
cyr["cum"] = cyr.sort_values("year_added").groupby("primary_country")["n"].cumsum()
top5c = cyr.query("year_added==2021").nlargest(5,"cum")["primary_country"].tolist()
race = cyr.query("primary_country in @top5c").copy()
race["rank"] = race.groupby("year_added")["cum"].rank(ascending=False, method="first")

fig, ax = plt.subplots(figsize=(10, 5.5))
for c in top5c:
    sub = race.query("primary_country==@c").sort_values("year_added")
    color = "#E53935" if c == "India" else "#DDDDDD"
    lw = 2.5 if c == "India" else 0.8
    ax.plot(sub.year_added, sub["rank"], "-o", color=color, lw=lw, markersize=4 if c=="India" else 2)
    ax.text(sub.year_added.values[-1]+0.15, sub["rank"].values[-1],
        f" {c}", fontsize=7, fontweight="bold" if c=="India" else "normal",
        color=color if c!="#DDDDDD" else "#AAA", va="center")
ax.set_ylim(5.5, 0.5); ax.set_xlim(2015.5, 2022.5)
ax.set_title("India rose from #5 to #2 in Netflix content production", fontsize=13)
ax.set_ylabel("Rank (1 = most titles)")
plt.tight_layout(); plt.savefig("output/slide06_deepdive.png", dpi=300); plt.close()

# S7: Seasonal — calendar heatmap
d2020 = nf.query("year_added==2020").dropna(subset=["date_added"]).groupby("date_added").size().reset_index(name="n")
d2020["dow"] = d2020["date_added"].dt.dayofweek
d2020["week"] = d2020["date_added"].dt.isocalendar().week.astype(int)
pivot = d2020.pivot_table(index="dow", columns="week", values="n", aggfunc="sum", fill_value=0)

fig, ax = plt.subplots(figsize=(11, 3.5))
im = ax.imshow(pivot.values, cmap="YlGn", aspect="auto", interpolation="nearest")
ax.set_yticks(range(7)); ax.set_yticklabels(["Mon","Tue","Wed","Thu","Fri","Sat","Sun"], fontsize=7)
ax.set_xlabel("Week of Year"); plt.colorbar(im, ax=ax, shrink=0.5, label="Titles")
ax.set_title("Friday is the primary release day, with Q4 holiday spikes", fontsize=12)
plt.tight_layout(); plt.savefig("output/slide07_seasonal.png", dpi=300); plt.close()

# S8: Resolution — composition stacked area
type_m = (nf.dropna(subset=["date_added"]).query("date_added.dt.year>=2015")
    .set_index("date_added").groupby("type").resample("ME").size().unstack(0, fill_value=0).cumsum())
total = type_m.sum(axis=1)
fig, ax = plt.subplots(figsize=(10, 5.5))
if "Movie" in type_m and "TV Show" in type_m:
    ax.stackplot(type_m.index, type_m["Movie"]/total*100, type_m["TV Show"]/total*100,
        labels=["Movie","TV Show"], colors=["#1565C0","#E53935"], alpha=0.7)
ax.legend(fontsize=8, loc="center right")
ax.set_title("TV Show share grew from 20% to 35% — the catalog is rebalancing", fontsize=12)
ax.set_ylabel("Share (%)"); ax.xaxis.set_major_formatter(DateFormatter("%Y"))
plt.tight_layout(); plt.savefig("output/slide08_resolution.png", dpi=300); plt.close()

# ═══════════════════════════════════════════════════════
# TITLE SEQUENCE TEST
# ═══════════════════════════════════════════════════════
print("\n=== TITLE SEQUENCE TEST ===")
titles = [
    "1. Netflix Content Strategy: A Data Story",
    "2. Netflix built a catalog of 8,800+ titles over a decade",
    f"3. Movie additions declined {decline}% from their 2019 peak",
    "4. TV Shows grew 3x faster than Movies in relative terms",
    f"5. {fastest_genre} is the fastest-growing genre category",
    "6. India rose from #5 to #2 in content production",
    "7. Friday is the primary release day, with Q4 spikes",
    "8. TV Show share grew from 20% to 35% — rebalancing",
    "9. Four actions: originals, local content, reduce filler",
    "10. Explore the dashboard for more detail",
]
for t in titles: print(f"  {t}")
print("\nDoes the story work from titles alone? ✓")

print("\nAll W07-M07 Python outputs saved")

File: latest/week3/workshop7_m7/W07_M07_demo.R
# ============================================================
# Workshop 7 — Module 7: Case Study — Netflix Data Story
# Complete 10-slide assertion-evidence deck in R
# ============================================================
library(tidyverse); library(patchwork); dir.create("output", showWarnings = FALSE)

# ── CLEAN THEME (from M06) ──────────────────────────────
theme_story <- theme_minimal(base_size = 12) +
  theme(panel.grid = element_blank(),
    axis.line = element_line(color = "#EEEEEE", linewidth = 0.3),
    plot.title = element_text(face = "bold", size = 14),
    plot.subtitle = element_text(color = "#666", size = 10),
    plot.caption = element_text(color = "#AAA", size = 7),
    legend.position = "none")
theme_set(theme_story)

# ── LOAD + PREPARE ──────────────────────────────────────
nf <- read_csv("netflix.csv") |>
  mutate(date_added = mdy(str_trim(date_added)),
    year_added = year(date_added),
    month_added = month(date_added),
    primary_country = str_trim(str_extract(country, "^[^,]+")))
nf_genres <- nf |> separate_rows(listed_in, sep = ", ") |> mutate(genre = str_trim(listed_in))

yearly_type <- nf |> filter(!is.na(year_added), year_added >= 2015, year_added <= 2021) |>
  count(year_added, type)
movies <- yearly_type |> filter(type == "Movie")
tv <- yearly_type |> filter(type == "TV Show")
peak <- movies |> slice_max(n, n = 1)
decline <- round((1 - movies$n[movies$year_added == 2021] / peak$n) * 100)

# ═══════════════════════════════════════════════════════
# SLIDE 2: CONTEXT — Cumulative catalog growth
# Story type: Zoom Out | Chart: Area
# ═══════════════════════════════════════════════════════
p2 <- nf |> filter(!is.na(date_added)) |>
  mutate(month = floor_date(date_added, "month")) |>
  count(month) |> mutate(cum = cumsum(n)) |>
  ggplot(aes(x = month, y = cum)) +
  geom_area(fill = "#1565C0", alpha = 0.15) +
  geom_line(color = "#1565C0", linewidth = 1.2) +
  scale_x_date(date_breaks = "2 years", date_labels = "%Y") +
  labs(title = "Netflix built a catalog of 8,800+ titles over a decade",
       subtitle = "Cumulative titles added to the platform (2008–2021)",
       x = NULL, y = "Cumulative Titles",
       caption = "Source: Netflix dataset | UNYT")
ggsave("output/slide02_context.png", p2, width = 10, height = 5.5, dpi = 300)

# ═══════════════════════════════════════════════════════
# SLIDE 3: COMPLICATION — Movie decline
# Story type: Change Over Time | Chart: Line (grey+accent)
# ═══════════════════════════════════════════════════════
p3 <- ggplot(yearly_type, aes(x = year_added, y = n, group = type)) +
  geom_line(color = "#DDDDDD", linewidth = 0.8) +
  geom_line(data = movies, color = "#1565C0", linewidth = 2.5) +
  geom_point(data = movies, color = "#1565C0", size = 3) +
  annotate("label", x = 2018.3, y = movies$n[movies$year_added == 2021] * 0.82,
    label = paste0("–", decline, "% from peak"),
    size = 3.5, fontface = "bold", color = "#E53935", fill = "white", label.size = 0.4) +
  geom_text(data = yearly_type |> filter(year_added == 2021),
    aes(label = type, color = type == "Movie"),
    hjust = -0.1, size = 3.5, fontface = "bold", show.legend = FALSE) +
  scale_color_manual(values = c("TRUE" = "#1565C0", "FALSE" = "#BBB")) +
  coord_cartesian(xlim = c(2015, 2022.5)) +
  labs(title = paste0("Movie additions declined ", decline, "% from their 2019 peak"),
       subtitle = "The largest content category is contracting",
       x = NULL, y = "Titles Added")
ggsave("output/slide03_complication.png", p3, width = 10, height = 5.5, dpi = 300)

# ═══════════════════════════════════════════════════════
# SLIDE 4: CONTRAST — Movie vs TV divergence
# Story type: Contrast | Chart: Dual line indexed
# ═══════════════════════════════════════════════════════
indexed <- yearly_type |>
  group_by(type) |> mutate(idx = n / first(n) * 100) |> ungroup()
p4 <- ggplot(indexed, aes(x = year_added, y = idx, color = type)) +
  geom_line(linewidth = 1.5) + geom_point(size = 2.5) +
  geom_hline(yintercept = 100, linetype = "dotted", color = "#888") +
  scale_color_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  geom_text(data = indexed |> filter(year_added == 2021),
    aes(label = type), hjust = -0.1, size = 3, fontface = "bold", show.legend = FALSE) +
  coord_cartesian(xlim = c(2015, 2022.5)) +
  labs(title = "TV Shows grew 3× faster than Movies in relative terms",
       subtitle = "Both indexed to 100 at 2015 for fair comparison",
       x = NULL, y = "Index (2015 = 100)")
ggsave("output/slide04_contrast.png", p4, width = 10, height = 5.5, dpi = 300)

# ═══════════════════════════════════════════════════════
# SLIDE 5: ANALYSIS — Genre growth (climax)
# Story type: Change | Chart: Bar (accent)
# ═══════════════════════════════════════════════════════
genre_growth <- nf_genres |>
  filter(!is.na(year_added), year_added %in% c(2017, 2021)) |>
  count(year_added, genre) |>
  pivot_wider(names_from = year_added, values_from = n, values_fill = 0) |>
  mutate(growth = (`2021` - `2017`) / pmax(`2017`, 1) * 100) |>
  slice_max(growth, n = 7)
fastest <- genre_growth |> slice_max(growth, n = 1) |> pull(genre)

p5 <- ggplot(genre_growth,
  aes(x = fct_reorder(genre, growth), y = growth,
      fill = genre == fastest)) +
  geom_col(width = 0.5, show.legend = FALSE) +
  geom_text(aes(label = paste0("+", round(growth), "%")),
    hjust = -0.1, size = 3, fontface = "bold") +
  scale_fill_manual(values = c("TRUE" = "#2E7D32", "FALSE" = "#BBDEFB")) +
  coord_flip(clip = "off") +
  labs(title = paste0(fastest, " is the fastest-growing genre category"),
       subtitle = "Genre growth rates 2017 → 2021 (green = fastest)",
       x = NULL, y = "Growth %")
ggsave("output/slide05_analysis.png", p5, width = 10, height = 5.5, dpi = 300)

# ═══════════════════════════════════════════════════════
# SLIDE 6: DEEP DIVE — India's rise
# Story type: Zoom Out | Chart: Bump
# ═══════════════════════════════════════════════════════
country_cum <- nf |>
  filter(!is.na(primary_country), !is.na(year_added), year_added >= 2016, year_added <= 2021) |>
  count(year_added, primary_country) |>
  group_by(primary_country) |> arrange(year_added) |> mutate(cum = cumsum(n)) |> ungroup()
top5c <- country_cum |> filter(year_added == 2021) |> slice_max(cum, n = 5) |> pull(primary_country)
race <- country_cum |> filter(primary_country %in% top5c) |>
  group_by(year_added) |> mutate(rank = rank(-cum, ties.method = "first")) |> ungroup()

p6 <- ggplot(race, aes(x = year_added, y = rank, group = primary_country)) +
  geom_line(color = "#DDDDDD", linewidth = 0.8) +
  geom_line(data = race |> filter(primary_country == "India"),
    color = "#E53935", linewidth = 2.5) +
  geom_point(data = race |> filter(primary_country == "India"),
    color = "#E53935", size = 3) +
  geom_text(data = race |> filter(year_added == 2021),
    aes(label = primary_country, color = primary_country == "India"),
    hjust = -0.15, size = 3, fontface = "bold", show.legend = FALSE) +
  scale_color_manual(values = c("TRUE" = "#E53935", "FALSE" = "#AAA")) +
  scale_y_reverse(breaks = 1:5) +
  coord_cartesian(xlim = c(2016, 2022.5)) +
  labs(title = "India rose from #5 to #2 in Netflix content production",
       subtitle = "Bump chart of top 5 content-producing countries (2016–2021)",
       x = NULL, y = "Rank (1 = most titles)")
ggsave("output/slide06_deepdive.png", p6, width = 10, height = 5.5, dpi = 300)

# ═══════════════════════════════════════════════════════
# SLIDE 7: SEASONAL — Release calendar
# Story type: Drill Down | Chart: Calendar heatmap
# ═══════════════════════════════════════════════════════
daily_2020 <- nf |> filter(year_added == 2020, !is.na(date_added)) |>
  count(date_added) |>
  mutate(dow = wday(date_added, label = TRUE, week_start = 1),
    week = isoweek(date_added))

p7 <- ggplot(daily_2020, aes(x = week, y = dow, fill = n)) +
  geom_tile(color = "white", linewidth = 0.4) +
  scale_fill_gradient(low = "#E8F5E9", high = "#1B5E20", name = "Titles") +
  scale_y_discrete(limits = rev(c("Mon","Tue","Wed","Thu","Fri","Sat","Sun"))) +
  theme(panel.grid = element_blank(), legend.key.size = unit(0.4, "cm")) +
  labs(title = "Friday is Netflix's primary release day, with Q4 holiday spikes",
       subtitle = "Calendar heatmap of daily additions in 2020 (row = DOW, col = week)",
       x = "Week of Year", y = NULL)
ggsave("output/slide07_seasonal.png", p7, width = 11, height = 4, dpi = 300)

# ═══════════════════════════════════════════════════════
# SLIDE 8: RESOLUTION — Quality > Quantity
# Story type: Change | Chart: Stacked area (composition shift)
# ═══════════════════════════════════════════════════════
type_monthly <- nf |> filter(!is.na(date_added), year(date_added) >= 2015) |>
  mutate(month = floor_date(date_added, "month")) |>
  count(month, type) |>
  group_by(type) |> arrange(month) |> mutate(cum = cumsum(n)) |> ungroup() |>
  group_by(month) |> mutate(pct = cum / sum(cum) * 100) |> ungroup()

p8 <- ggplot(type_monthly, aes(x = month, y = pct, fill = type)) +
  geom_area(alpha = 0.7, color = "white", linewidth = 0.3) +
  scale_fill_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  geom_text(data = type_monthly |> filter(month == max(month)),
    aes(label = paste0(type, " ", round(pct), "%")),
    position = position_stack(vjust = 0.5), size = 3, fontface = "bold", color = "white") +
  scale_x_date(date_breaks = "1 year", date_labels = "%Y") +
  labs(title = "TV Show share grew from 20% to 35% — the catalog is rebalancing",
       subtitle = "Proportional composition of cumulative Netflix catalog over time",
       x = NULL, y = "Share (%)")
ggsave("output/slide08_resolution.png", p8, width = 10, height = 5.5, dpi = 300)

# ═══════════════════════════════════════════════════════
# COMPOSE: Full 8-slide panel (slides 2–9 as evidence)
# ═══════════════════════════════════════════════════════
all_slides <- (p2 + labs(title = "2. Context") + theme(plot.title = element_text(size = 9))) |
  (p3 + labs(title = "3. Complication") + theme(plot.title = element_text(size = 9)))
all_slides2 <- (p4 + labs(title = "4. Contrast") + theme(plot.title = element_text(size = 9))) |
  (p5 + labs(title = "5. Analysis") + theme(plot.title = element_text(size = 9)))
all_slides3 <- (p6 + labs(title = "6. Deep Dive") + theme(plot.title = element_text(size = 9))) |
  (p7 + labs(title = "7. Seasonal") + theme(plot.title = element_text(size = 9)))
all_slides4 <- (p8 + labs(title = "8. Resolution") + theme(plot.title = element_text(size = 9))) |
  plot_spacer()

ggsave("output/netflix_story_deck.png",
  all_slides / all_slides2 / all_slides3 / all_slides4 +
  plot_annotation(title = "Netflix Data Story: 8-Slide Evidence Deck"),
  width = 16, height = 22, dpi = 200)

# ── TITLE SEQUENCE TEST ──────────────────────────────────
cat("\n── TITLE SEQUENCE TEST ──\n")
cat("Read only the titles. Does the story work?\n\n")
cat("1. Netflix Content Strategy: A Data Story\n")
cat("2. Netflix built a catalog of 8,800+ titles over a decade\n")
cat("3. Movie additions declined", decline, "% from their 2019 peak\n")
cat("4. TV Shows grew 3× faster than Movies in relative terms\n")
cat("5.", fastest, "is the fastest-growing genre category\n")
cat("6. India rose from #5 to #2 in content production\n")
cat("7. Friday is Netflix's primary release day, with Q4 spikes\n")
cat("8. TV Show share grew from 20% to 35% — the catalog is rebalancing\n")
cat("9. Four actions: originals, local content, reduce filler, language hubs\n")
cat("10. Explore the interactive dashboard for more detail\n")

cat("\n── All W07-M07 R outputs saved ──\n")
